src/libraries/CalldataDecoder.sol

```diff
@@ -228,38 +254,29 @@ library CalldataDecoder {
         }
     }
 
-    /// @notice Decode the `_arg`-th element in `_bytes` as a dynamic array
-    /// @dev The decoding of `length` and `offset` is universal,
-    /// whereas the type declaration of `res` instructs the compiler how to read it.
-    /// @param _bytes The input bytes string to slice
-    /// @param _arg The index of the argument to extract
-    /// @return length Length of the array
-    /// @return offset Pointer to the data part of the array
-    function toLengthOffset(bytes calldata _bytes, uint256 _arg)
-        internal
-        pure
-        returns (uint256 length, uint256 offset)
-    {
-        uint256 relativeOffset;
-        assembly ("memory-safe") {
-            // The offset of the `_arg`-th element is `32 * arg`, which stores the offset of the length pointer.
-            // shl(5, x) is equivalent to mul(32, x)
-            let lengthPtr := add(_bytes.offset, calldataload(add(_bytes.offset, shl(5, _arg))))
-            length := calldataload(lengthPtr)
-            offset := add(lengthPtr, 0x20)
-            relativeOffset := sub(offset, _bytes.offset)
-        }
-        if (_bytes.length < length + relativeOffset) revert SliceOutOfBounds();
-    }
-
     /// @notice Decode the `_arg`-th element in `_bytes` as `bytes`
     /// @param _bytes The input bytes string to extract a bytes string from
     /// @param _arg The index of the argument to extract
     function toBytes(bytes calldata _bytes, uint256 _arg) internal pure returns (bytes calldata res) {
-        (uint256 length, uint256 offset) = toLengthOffset(_bytes, _arg);
+        uint256 length;
         assembly ("memory-safe") {
+            // The offset of the `_arg`-th element is `32 * arg`, which stores the offset of the length pointer.
+            // shl(5, x) is equivalent to mul(32, x)
+            let lengthPtr :=
+                add(_bytes.offset, and(calldataload(add(_bytes.offset, shl(5, _arg))), OFFSET_OR_LENGTH_MASK))
+            // the number of bytes in the bytes string
+            length := and(calldataload(lengthPtr), OFFSET_OR_LENGTH_MASK)
```

Comment:
Here the `length` value is silently trimmed to 32 bytes, which could hide errors in abi encoding.  Consider reverting in case the `length` value exceeds 32 bits.